src/server/db_slice.cc

```diff
@@ -481,6 +482,13 @@ OpResult<DbSlice::PrimeItAndExp> DbSlice::FindInternal(const Context& cntx, std:
 
   if (caching_mode_ && IsValid(res.it)) {
     if (!change_cb_.empty()) {
+      {
+        // We need this because we might not be able to acquire this lock within
+        // CallChangeCallbacks and if this happens then Capture Version Change
+        // might be invalid, since the bucket might change after we block waiting
+        // for the lock to be available.
+        std::unique_lock lk(local_mu_);
```

Comment:
I don't understand - why release the lock before calling `CVCUponBump()`?
And why can't we acquire this lock inside it?

Reply:
Because `CVCUponBump()`  acquires the same lock (`CallChangeCallbacks` below does it) so it will deadlock. 

Furthermore, `CVCUponBump` captures the version change meaning that it `logically tries to insert` an element. Logically, meaning that it does not perform any insertion but it calculates which buckets might change upon the insertion and we serialize those buckets via the callbacks when the bucket version is less than the one registered in the callback. Then we can safely make the change.

The problem lies within `CVCUponBump` which first makes this logical calculation above and then it applies the callbacks. However, the callbacks might fail to acquire the lock because someone else might already have acquired this lock and by the time we wake up we might have the wrong calculation (since some other fiber might wake up, do the same calculation and bump up the version). 

By locking above, we guarantee than when we wake up, we will acquire the lock again within the callback.

Reply:
This seems fragile...
What if one of the callbacks preempt, and the segment changes structure before the next callbacks are called? It could then be called with invalid parameters, no?
I think that a more robust solution would be to mark `CallChangeCallbacks()` as `EXCLUSIVE_LOCKS_REQUIRED` (which means the mutex must be held when calling it), and then lock it before each of the calls. Here you'll hold the mutex locked and not release it, thus you won't deadlock.
wdyt?

Reply:
> What if one of the callbacks preempt, and the segment changes structure before the next callbacks are called? It could then be called with invalid parameters, no?

This can't happen because we hold the lock when we preempt.

> I think that a more robust solution would be to mark CallChangeCallbacks() as EXCLUSIVE_LOCKS_REQUIRED 

IMO that's what I had in mind so I couldn;t agree more :100:

Reply:
> This can't happen because we hold the lock when we preempt.
> 

But we don't hold it throughout the preemptions, right? We hold it in each preemption within each callback, but there could be multiple.
Anyway, looks like we agree on a solution :)

Reply:
Yes but the whole block is atomic implicitly, once the lock is released after the preemption it will be acquired by the same fiber until it applies all the callbacks (which each one might preempt). The reason is that upon releasing the lock, the fiber is not put to sleep but rather it continues and it just calls the next callback which immediately reacquires the lock(when we release the lock we just update the waiting queue, but this merely wakes up a fiber and activates it on the scheduler -- we don't switch to another fiber upon release)

Reply:
I agree that we won't switch fibers when we unlock here, until the first preemption. But the first preemption will switch fibers, at which point we're in the danger zone

Reply:
but we hold the lock when we preempt no ? What am I missing here?